classes/migration/upgrade/v3_4_0/MergeLocalesMigration.php

```diff
@@ -0,0 +1,366 @@
+            throw new Exception('Upgrade could not be completed because required properties for the MergeLocalesMigration migration are undefined.');
+        }
+
+        // All _settings tables.
+        $settingsTables = $this->getSettingsTables();
+        foreach($settingsTables as $settingsTable) {
+            if (Schema::hasColumn($settingsTable, 'locale')) {
+                $affectedLocales = $this->getAffectedLocales();
+                foreach($affectedLocales as $affectedLocale => $toLocale) {
+                    if (!isset($toLocale)) {
+                        DB::table($settingsTable)
+                            ->where('locale', 'like', $affectedLocale . '_%')
+                            ->update(['locale' => $affectedLocale]);
+                    } else {
+                        DB::table($settingsTable)
+                            ->where('locale', 'like', $affectedLocale)
+                            ->update(['locale' => $toLocale]);
+                    }
+                    
+                }
+            }
+        }
+
+        // Tables
+        // site
+        $site = DB::table('site')
+            ->select(['supported_locales', 'installed_locales', 'primary_locale'])
+            ->first();
+
+        $this->updateArrayLocaleNoId($site->supported_locales, 'site', 'supported_locales');
+        $this->updateArrayLocaleNoId($site->installed_locales, 'site', 'installed_locales');
+        $this->updateSingleValueLocaleNoId($site->primary_locale, 'site', 'primary_locale');
+        
+        // users
+        $users = DB::table('users')
+            ->get();
+
+        foreach ($users as $user) {
+            $this->updateArrayLocale($user->locales, 'users', 'locales', 'user_id', $user->user_id);
+        }
+
+        // submissions
+        $submissions = DB::table('submissions')
+            ->get();
+
+        foreach ($submissions as $submission) {
+            $this->updateSingleValueLocale($submission->locale, 'submissions', 'locale', 'submission_id', $submission->submission_id);
+        }
+
+        // email_templates_default_data
+        $emailTemplatesDefaultData = DB::table('email_templates_default_data')
+            ->get();
+
+        foreach ($emailTemplatesDefaultData as $emailTemplatesDefaultDataCurrent) {
+            $this->updateSingleValueLocaleEmailData($emailTemplatesDefaultDataCurrent->locale, 'email_templates_default_data', 'locale', 'email_key', $emailTemplatesDefaultDataCurrent->email_key);
+        }
+
+        // Context
+        $contexts = DB::table($this->CONTEXT_TABLE)
+            ->get();
+
+        foreach ($contexts as $context) {
+            $this->updateSingleValueLocale($context->primary_locale, $this->CONTEXT_TABLE, 'primary_locale', $this->CONTEXT_COLUMN, $context->{$this->CONTEXT_COLUMN});
+        }
+
+        $journalSettingsFormLocales = DB::table($this->CONTEXT_SETTINGS_TABLE)
```

Comment:
OJS-specific language ("journal") shouldn't be used in pkp-lib; use application-independent "context".